library/std/src/sys/thread_local/fast_local.rs

```diff
@@ -13,22 +13,21 @@ pub macro thread_local_inner {
     (@key $t:ty, const $init:expr) => {{
         #[inline]
         #[deny(unsafe_op_in_unsafe_fn)]
-        // FIXME: Use `SyncUnsafeCell` instead of allowing `static_mut_refs` lint
-        #[cfg_attr(bootstrap, allow(static_mut_ref))]
-        #[cfg_attr(not(bootstrap), allow(static_mut_refs))]
         unsafe fn __getit(
             _init: $crate::option::Option<&mut $crate::option::Option<$t>>,
         ) -> $crate::option::Option<&'static $t> {
             const INIT_EXPR: $t = $init;
             // If the platform has support for `#[thread_local]`, use it.
             #[thread_local]
-            static mut VAL: $t = INIT_EXPR;
+            // We use `UnsafeCell` here instead of `static mut` to ensure any generated TLS shims
+            // have a nonnull attribute on their return value.
```

Comment:
Could you share an example of this? For me, `nonnull` [is emitted](https://godbolt.org/z/jo3dbnexb) even with `static mut`, but perhaps I'm misunderstanding your meaning here?

Reply:
That may be LLVM inferring it, which won't be the case from other crates without LTO.

Reply:
How is that changed by `UnsafeCell`?

Reply:
It's not. It's changed by `static` resulting in the TLS shim returning a reference (which are always nonnull) vs `static mut` where the TLS shim returns a raw pointer.

Reply:
Ah yes, I [see](https://github.com/rust-lang/rust/blob/master/compiler/rustc_middle/src/ty/util.rs#L652) now!